src/python/zquantum/core/wip/circuit/gates/_gate.py

```diff
@@ -0,0 +1,388 @@
+"""Base classes for implementing quantum gates."""
+from abc import abstractmethod, ABC
+import inspect
+import numpy as np
+import sympy
+import json
+import copy
+import warnings
+from typing import Tuple, Union, Dict, TextIO, Set, Any, Optional
+from ....utils import SCHEMA_VERSION
+
+
+def _evaluate_parameter(param, symbol_map):
+    if isinstance(param, sympy.Symbol):
+        return symbol_map.get(param, param)
+    elif isinstance(param, sympy.Expr):
+        return param.subs(symbol_map)
+    else:
+        return param
+
+
+class Gate(ABC):
+    """Base class for quantum Gates.
+
+    Args:
+        qubits:  tuple of qubit indices this gates acts on.
+
+    Attributes:
+        qubits:  tuple of qubit indices this gates acts on.
+    """
+
+    def __init__(self, qubits: Tuple[int, ...]):
+        if not self.are_qubits_unique(qubits):
+            raise ValueError("Qubits need to be unique.")
+        self.qubits = qubits
+
+    @property
+    @abstractmethod
+    def matrix(self) -> sympy.Matrix:
+        """Matrix representation of this gate."""
+        ...
+
+    @property
+    def params(self) -> Tuple[Any, ...]:
+        return ()
+
+    @property
+    def symbolic_params(self) -> Set[str]:
+        """Free symbols present in gate's matrix."""
+        return set(
```

Comment:
Again, @mstechly where did we land on this?

Reply:
I don't understand the context really :(

Reply:
Oh, we landed on returning a sorted list:

```
        all_symbols = []
        for param in self.params:
            if isinstance(param, sympy.Basic):
                for symbol in param.free_symbols:
                    all_symbols.append(symbol)

        unique_symbols = sorted(list(set(all_symbols)), key=str)
        return unique_symbols
```

Reply:
See here: https://github.com/zapatacomputing/z-quantum-core/blob/6449cc6e3a43d9b44ef60682458111f628dc2ba1/src/python/zquantum/core/circuit/_gate.py#L120

Reply:
Added the param sorting in 97abcdc

Reply:
Just fix the return type of the function please.

Reply:
✅ e6d1789